shared/windows/scandir.c

```diff
@@ -1,86 +1,153 @@
-#include <dirent.h>
 #include <stdlib.h>
-#include <string.h>
+#include <dirent.h>
+#include <errno.h>
 #include <windows.h>
-#include <stdio.h>
-#include "utils.h"
 
-#define DIRENT_CHUNK 64
-// TODO: no selector or cmp support
+//#include "common.h" // for trace_err()
+
+static void set_convert_errno() {
+    // Is there like a debug assert instead in this project?
+    /*if (GetLastError() != ERROR_NO_UNICODE_TRANSLATION) {
+        // Sad days =(
+        trace_err(
+            "The scandir implementation is using a bad argument for encoding conversion!\n"
+        );
+    }*/
+    _set_errno(EILSEQ);
+}
+
+static void free_namelist(struct dirent **names, int count) {
+    for (int i = 0; i < count; i++) {
+        free(names[i]);
+    }
+    free(names);
+}
+
 int scandir (const char      *dirname_o,
              struct dirent ***namelist_to,
              int            (*selector) (const struct dirent *),
              int            (*cmp) (const struct dirent **, const struct dirent **)) {
+    if (dirname_o == NULL) {
+        return -1;
+    }
 
-    struct dirent ** namelist = malloc (sizeof(struct dirent *) * DIRENT_CHUNK);
-    if (!namelist) {
+    int sErr = errno; // Don't disturb errno unless there is an error.
+
+    // setting arg cchWideChar to zero will return the length in codepoints, and
+    // it also includes str terminator in the length
+    int utf16_points = MultiByteToWideChar(
+        CP_UTF8, MB_ERR_INVALID_CHARS, dirname_o, -1, NULL, 0
+    );
+    if (utf16_points < 1) {
+        set_convert_errno();
         return -1;
     }
 
+    // Add 4 for enough room for '\*.*'
+    wchar_t dir_wide[utf16_points + 4];
+    memset(dir_wide, 0, sizeof(wchar_t) * (utf16_points + 4));
+    int ret = MultiByteToWideChar(
+        CP_UTF8, MB_ERR_INVALID_CHARS, dirname_o, -1, dir_wide, utf16_points
```

Comment:
There is no need to wrap this.
```int utf16_points = MultiByteToWideChar(CP_UTF8, MB_ERR_INVALID_CHARS, dirname_o, -1, NULL, 0);```

Function headers are ok though.

Reply:
Understood, I have fixed all the wrapping. The following commit fix the wrapping as requested.

commit: c218e31c2d7095847fad049237ed9dd0ef531712